### PR TITLE
docs: improve 'insertmode' emulation

### PR DESCRIPTION
**Pull request by zeertzjq**
The current emulation script enters Insert mode much too frequently.
Using only BufWinEnter seems to be a closer simulation.

Also add a few more mappings.

### CODE REVIEW

**Comment by zeertzjq** [comment]
cmdwin and Insert mode abbreviation are a bit hard to deal with. I'll leave them alone.

**Comment by zeertzjq** [comment]
```vim
autocmd BufWinEnter * silent! if &modifiable | startinsert | endif
inoremap <Esc> <C-X><C-Z>
inoremap <C-C> <C-X><C-Z>
inoremap <C-L> <Esc>
inoremap <C-Z> <Cmd>stop<CR>
noremap <Esc> 1<Esc><Cmd>startinsert<CR>
noremap <C-C> 1<C-C><Cmd>startinsert<CR>
noremap <C-\><C-G> <C-\><C-N><Cmd>startinsert<CR>
cnoremap <C-\><C-G> <C-\><C-N><Cmd>startinsert<CR>
inoremap <C-\><C-G> <C-X><C-Z>
```

@M1cha @frabjous Does this script work well?

**Review by justinmk — APPROVED** [review]

**Comment by zeertzjq** [comment]
> if I'm in selectmode because of having used the shift keys (keymodel=startsel,stopsel; selectmode=mouse,key) and leave it with an arrow (up arrow, down arrow), at least some times I end up in normal mode rather than back in insert.

Hmm, I tried doing this, and even without this script, doing these things ends in Insert mode.

File: runtime/doc/vim_diff.txt [hunk 1]
```diff
@@ -550,10 +550,16 @@ Options:
...
 >
-      autocmd VimEnter,CmdlineLeave,WinEnter,WinScrolled,BufEnter * silent! if &modifiable | startinsert | endif
-      inoremap <Esc> <Nop>
-      inoremap <C-L> <Esc>
-      nnoremap <Esc> i
+      autocmd BufWinEnter * silent! if &modifiable | startinsert | endif
+      inoremap <Esc> <C-X><C-Z><C-]>
+      inoremap <C-C> <C-X><C-Z>
+      inoremap <C-L> <C-X><C-Z><C-]><Esc>
+      inoremap <C-Z> <C-X><C-Z><Cmd>stop<CR>
```
Comment (justinmk) [inline runtime/doc/vim_diff.txt:557]: nit: in docs and scripts it's good practice to spell out the full command name.
```suggestion
      inoremap <C-Z> <C-X><C-Z><Cmd>stopinsert<CR>
```

**Comment by zeertzjq** [comment]
Hmm, this may require timers.

**Comment by justinmk** [comment]
If this improves the current script , let's merge it so that we don't forget.

**Comment by zeertzjq** [comment]
I've notice that if `<Esc>` comes from a `noremap` mapping then it is not possible fully simulate its behavior with `'insertmode'` using mappings. In this case `vim.on_key()` may be needed.

**Comment by frabjous** [comment]
Not sure if this will be helpful; maybe this is all understood already …

But I've been trying to experiment with the proposed script with a fairly minimal-ish config file.

(I have a lot in the config I actually use which makes it harder to narrow down the source of complications, so I'm trying to avoid it, which is tough given how used to it I am.)

Here are some ways in which the proposal doesn't replicate insertmode.

I've been able to reliably recreate the problem I mentioned in my previous post using this fairly minimal config:

```
autocmd BufWinEnter * silent! if &modifiable | startinsert | endif
inoremap <Esc> <C-X><C-Z><C-]>
inoremap <C-C> <C-X><C-Z>
inoremap <C-L> <C-X><C-Z><C-]><Esc>
inoremap <C-Z> <C-X><C-Z><Cmd>suspend<CR>
noremap <Esc> 1<Esc><Cmd>startinsert<CR>
noremap <C-C> 1<C-C><Cmd>startinsert<CR>
noremap <C-\><C-G> <C-\><C-N><Cmd>startinsert<CR>
cnoremap <C-\><C-G> <C-\><C-N><Cmd>startinsert<CR>
inoremap <C-\><C-G> <C-X><C-Z>

set selectmode=mouse,key
set keymodel=startsel,stopsel
set wrap
inoremap <Up> <C-o>gk
inoremap <Down> <C-o>gj
vnoremap <Up> <C-c>gkgV
vnoremap <Down> <C-c>gjgV
```
I think settings similar to these would be pretty common for a lot of people. If you have wrap on you typically don't want the up and down arrow keys to navigate by file line, but by wrapped lines.

Now in some wrapped text, start select mode with shift-up or shift-down and use shift-up/down to select more. Now press up or down without shift. The cursor will move, you'll leave select mode (as you should with these settings), but you'll be left in normal mode, not insert mode.

If I simplify the last two keymaps to `<C-c>gk`/`<C-c>gj`, then the up and down without shift don't move the cursor, don't take me out of select mode, but *do* take me out of "(insert) SELECT" into SELECT, and when I then move out with left or right arrows, I'm left in normal mode.

I am also having an issue if I use any mapping in selection mode that uses `<C-c>` to go into command mode to do something; doesn't matter much what it is.

Consider this mapping (just a silly example but it makes the point):

```
vnoremap <C-h> <C-c>:echo 'hi'<CR>
```
Using this takes me out of "(insert) SELECT" into "SELECT", and when I exit select mode (however I do, e.g., clicking the mouse elsewhere or using the arrows), I am put into normal mode.

If I add `gV` at the end of the mapping, I am put into normal mode immediately when pressing `<C-h>`.

These both seem to have something to do with the difference between what's marked at the bottom as "SELECT" or "(insert) SELECT", but I'm having trouble finding documentation about that difference. Maybe I could use more elaborate bindings if I understood that difference; not sure.

The only thing that seems to solve both these problems is something quite extreme like `autocmd ModeChanged *:n startinsert`.

**Comment by zeertzjq** [comment]
I have an updated script:
```vim
autocmd BufWinEnter * silent! if &modifiable | startinsert | endif
inoremap <Esc> <C-X><C-Z><C-]>
inoremap <expr> <C-C> getcmdwintype() !=# '' ? '<C-C>' : '<C-X><C-Z>'
inoremap <C-L> <C-X><C-Z><C-]><Esc>
inoremap <C-Z> <C-X><C-Z><Cmd>suspend<CR>
noremap <expr> <C-C> getcmdwintype() !=# '' ? '<C-C>' : '<Esc>'
snoremap <expr> <C-C> getcmdwintype() !=# '' ? '<C-C>' : '<Esc>'
noremap <C-\><C-G> <C-\><C-N><Cmd>startinsert<CR>
cnoremap <C-\><C-G> <C-\><C-N><Cmd>startinsert<CR>
inoremap <C-\><C-G> <C-X><C-Z>

lua << EOF
  vim.on_key(function(c)
    if c == '\003' or c == '\027' then
      local mode = vim.api.nvim_get_mode().mode
      if mode:find('^[nvV\022sS\019]') and vim.fn.getcmdwintype() == '' then
        vim.schedule(function() vim.cmd('startinsert') end)
      end
    end
  end, 0)
EOF
```

**Comment by zeertzjq** [comment]
A problem here is that when using this script, showing a hidden buffer in a window using `nvim_win_set_buf` also enters Insert mode, while with `'insertmode'` this only happens when using `:edit`.

**Comment by frabjous** [comment]
The newer version does seem better regarding the issues I described. I'll test it for awhile and let you know if I come across anything else. Thanks.

**Review by justinmk — APPROVED** [review]
on_key looks risky but this is an improvement. Thanks @zeertzjq

**Comment by zeertzjq** [comment]
This part is the hardest part to simulate:
```c
	// Don't reset "restart_edit" when 'insertmode' is set, it won't be
	// set again below when halfway a mapping.
	if (!p_im)
	    restart_edit = 0;
```

Now `restart_edit` is always set to 0 when Ctrl-C is received in Normal mode, it is not possible to know whether Replace mode needs to be restarted when Ctrl-C is received halfway a mapping.

**Comment by zeertzjq** [comment]
> I am also having an issue if I use any mapping in selection mode that uses `<C-c>` to go into command mode to do something; doesn't matter much what it is.

It seems `<Esc>` can be used instead of `<C-C>` here. `<Esc>` should be better to stop Visual mode, as `<C-C>` can close cmdwin.

**Comment by zeertzjq** [comment]
I removed some handling for using `<C-C>` inside a mapping from the script, as `<C-C>` behavior is complicated, and it is better to use `<Esc>` inside a mapping instead.

**Pull request merged**